### PR TITLE
Fix device_allocator option in new benchmark suite

### PR DESCRIPTION
**Pull request by pzread**
Move `--device_allocator=caching` from tool to the place generates run flags (introduced in #11979). Also enable the option for e2e benchmark model tests.

The option is correctly enabled on the legacy benchmark suite in the original PR.

benchmarks: x86_64, cuda

### CODE REVIEW

**Review by antiagainst — APPROVED** [review]

File: build_tools/python/e2e_model_tests/run_module_utils.py
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/run_module_utils.py]: This feels a bit ad-hoc. Basically we're saying that we *always* run with the caching allocator. I don't think that we want to hardcode this. It should be part of the execution configuration.

Comment (pzread) [reply build_tools/python/e2e_model_tests/run_module_utils.py]: Yeah, just had the same thought. I decided to have a `global_flags` in the execution config for this kind of special case. Updated.

Comment (pzread) [reply build_tools/python/e2e_model_tests/run_module_utils.py]: Found some Python tricks to append the global flags to the extra flags. Not sure if it's easy to read...

Comment (pzread) [reply build_tools/python/e2e_model_tests/run_module_utils.py]: Ah, nvm, it doesn't allow us to reconstruct the dataclass from dict. Need to fine another solution

File: build_tools/python/e2e_test_framework/definitions/iree_definitions.py
Comment (GMNGeoffrey) [inline build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: This has confusing naming "extra" and "additional" mean the same thing, so it doesn't make clear what the distinction is between these (one is for initialization and the other is used). But I think more generally this is overly complex. Not so much because of the dataclasses cleverness, but rather because it makes this global flag implicit. What I was suggesting was that we really should make the usage of caching allocator part of the benchmark. It should be possible to use something else, and if we do then that should have a different benchmark ID. So the starting place I would use is just including it in "extra_flags" for every config. Ok, so that's annoying, let's add a function so we don't have to repeat ourselves so much. I think where that gets us is something like a helper function that appends "--device_allocator=caching" to extra_flags. I'd suggest this:


```suggestion
  extra_flags: List[str] = dataclasses.field(default_factory=list)
  
  @classmethod
  def with_caching_allocator(cls, *, id, tags, loader, driver, extra_flags) -> "ModuleExecutionConfig":
    extra_flags = extra_flags or []
    extra_flags.append("--device_allocator=caching")
    return cls(id, tags, loader, driver, extra_flags)
```

We could call the function `with_default_flags` as well, although then there's the danger that we could want to add a new default flag. This is all just syntactic sugar though, so we can play with the code later. I think explicit is better than implicit here though.

File: build_tools/python/e2e_model_tests/cmake_generator.py
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/cmake_generator.py]: Wat. This seems not great and I don't understand why we need to hardcode this

Comment (pzread) [reply build_tools/python/e2e_model_tests/cmake_generator.py]: It's because the CMake rule we use `iree_benchmark_suite_module_test` -> `iree_run_module_test` -> `iree_native_test` use a separate `DRIVER` arg for populating flags and test labels.

I rewrote to give `build_run_flags_for_execution_config` an option to generate flags without a driver and avoid hardcoded part here.

File: build_tools/python/e2e_test_framework/definitions/iree_definitions.py
Comment (pzread) [reply build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: SG.

It's a default flag only for benchmarking (at least from description in the original PR), so I put it in `benchmark_suites.iree.module_execution_configs` instead. If we ever want to enable it even broader, it probably should be the default in iree runtime.

Comment (GMNGeoffrey) [reply build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: It's *currently* the default for benchmarking, but the original PR mentions that we will hopefully have benchmarks in the future where it would be the wrong choice:


> For now this means that `--device_allocator=caching` needs to be passed to benchmark tooling when used as there's not yet a good way to force it on just for that binary (or if we'd even want to - hopefully benchmarks end up having some dynamic behavior in which case the caching allocator would be the wrong one to use). Clever users wanting to more tightly control the behavior can use a somewhat esoteric syntax to configure pools:

File: build_tools/python/benchmark_suites/iree/module_execution_configs.py
Comment (GMNGeoffrey) [inline build_tools/python/benchmark_suites/iree/module_execution_configs.py]: Don't use mutable default values: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

```suggestion
    extra_flags: Optional[List[str]] = None) -> iree_definitions.ModuleExecutionConfig:
    extra_flags = extra_flags or []
```

File: build_tools/python/e2e_model_tests/run_module_utils.py
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/run_module_utils.py]: It's generally clearer to use positive booleans to avoid confusing double-negatives. But also I think that arg isn't really necessary, as noted above

File: build_tools/python/e2e_model_tests/cmake_generator.py
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/cmake_generator.py]: Isn't this just

```suggestion
    runner_args += test_config.execution_config.extra_flags
```

File: build_tools/python/e2e_test_framework/definitions/iree_definitions.py
Comment (pzread) [reply build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: Then I think the right way is to have a "caching-allocator" tag on the current execution config, just to remind us all of these benchmarks are run with the caching allocator.

If we have other benchmarks w/o caching allocator, we can have the separate execution config for them (and potentially rename the current `ELF_LOCAL_SYNC_CONFIG` to `ELF_LOCAL_SYNC_CACHING_ALLOCATOR_CONFIG`). But I don't think we need to make that much noise for now (and we are free to change these names as long as the actually IDs stay the same).

File: build_tools/python/e2e_model_tests/cmake_generator.py
Comment (pzread) [reply build_tools/python/e2e_model_tests/cmake_generator.py]: I don't feel it's really future-proof to just use `extra_flags`, given we intend to generate some flags from the execution config besides the `extra_flags`. (And we will likely forget to update here if we generate other flags later).

This part will probably go way later if we fix the #11136

I have some idea about storing a full list of flags in these objects (and serializing them in the dump) instead of generating them everywhere. But that will be in the future changes.

File: build_tools/python/e2e_model_tests/run_module_utils.py
Comment (pzread) [reply build_tools/python/e2e_model_tests/run_module_utils.py]: Changed to the positive boolean.

File: build_tools/python/benchmark_suites/iree/module_execution_configs.py
Comment (pzread) [reply build_tools/python/benchmark_suites/iree/module_execution_configs.py]: Done.

File: build_tools/python/e2e_test_framework/definitions/iree_definitions.py
Comment (GMNGeoffrey) [reply build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: Do we really need to include it in tags? One of the advantages to artificial IDs is that we no longer need tags to provide permanent unique IDs. I wouldn't add this until we have usage of non-caching allocator. In fact, I think we could probably drop "default-flags" too (though let's wait on that)

File: build_tools/python/e2e_model_tests/run_module_utils.py
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/run_module_utils.py]: You're doing this to copy it, right? I think `list.copy` would be clearer

```suggestion
  run_flags = module_execution_config.extra_flags.copy()
```

https://docs.python.org/3/tutorial/datastructures.html#:~:text=list%20in%20place.-,list.copy(),-Return%20a%20shallow

**Review by GMNGeoffrey — APPROVED** [review]

**Comment by GMNGeoffrey** [comment]
Sorry, I disabled auto-merge with an accidental tap when I was trying to scroll on the mobile app

File: build_tools/python/e2e_model_tests/cmake_generator.py
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/cmake_generator.py:29]: Not really clear what the TODO is here. It seems like this is already done? What would you want to change?

**Review by GMNGeoffrey — APPROVED** [review]

File: build_tools/python/e2e_test_framework/definitions/iree_definitions.py
Comment (pzread) [reply build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: Removed

File: build_tools/python/e2e_model_tests/cmake_generator.py
Comment (pzread) [reply build_tools/python/e2e_model_tests/cmake_generator.py:29]: I still want to try to generate the flags for tests w/o have the separate DRIVER field. Some changes on the CMake rules are required.

**Comment by benvanik** [comment]
This didn't work - it disabled the caching allocator. The benchmarks posted above did report the regressions but it looks like it was merged anyway?
image
https://perf.iree.dev/serie?IREE?MobileNetV2%20%5Bfp32%2Cimagenet%5D%20(TFLite)%204-thread%2Cbig-core%2Cfull-inference%2Cdefault-flags%20with%20IREE-LLVM-CPU%20%40%20Pixel-4%20(CPU-ARMv8.2-A)

@pzread can you take a look at this asap?

**Comment by pzread** [comment]
> Abbreviated Benchmark Summary



> ## Abbreviated Benchmark Summary
> @ commit [3501968a62b30baac1537632745fcedba87d01aa](https://github.com/iree-org/iree/commit/3501968a62b30baac1537632745fcedba87d01aa) (vs. base [14828ecc5e7920706640da8dfc64cd6b419c7560](https://github.com/iree-org/iree/commit/14828ecc5e7920706640da8dfc64cd6b419c7560))
> 
> ### Regressed Latencies 🚩
> Benchmark Name	Average Latency (ms)	Median Latency (ms)	Latency Standard Deviation (ms)
> [MobileNetV3Small\_fp32 [fp32,imagenet] (exported\_tflite) [experimental-flags,fuse-padding][1-thread,full-inference,default-flags] with IREE-LLVM-CPU @ GCP-c2-standard-16 (CPU-x86\_64-CascadeLake)](https://perf.iree.dev/serie?IREE?c27738e97498c969076d1a2a693322821dd104dbcf7ba6e129ba893584bb0dfd)	5.038 (vs. 3.973, 26.80%↑)	5.027	0.049
> [MobileNetV3Small\_fp32 [fp32,imagenet] (exported\_tflite) [experimental-flags,fuse-padding][full-inference,default-flags] with IREE-LLVM-CPU-Sync @ GCP-c2-standard-16 (CPU-x86\_64-CascadeLake)](https://perf.iree.dev/serie?IREE?485da7a706b6c0940ef45626ec12ab149da295cc6a3c0a2c63e5a15a952580b4)	4.924 (vs. 3.907, 26.03%↑)	4.924	0.042
> [MobileNetV2\_fp32 [fp32,imagenet] (exported\_tflite) [experimental-flags,fuse-padding][1-thread,full-inference,default-flags] with IREE-LLVM-CPU @ GCP-c2-standard-16 (CPU-x86\_64-CascadeLake)](https://perf.iree.dev/serie?IREE?6272e089c33b7c5333b6188b6f61fbb15e7b6a0e9fcd9d54b3b7271cd730e0da)	12.894 (vs. 10.832, 19.04%↑)	12.886	0.040
> _[Top 3 out of 54 results showed]_
> 
> ### Regressed Compilation Times 🚩
> Benchmark Name	Compilation Time (ms)
> [MobileNetV3Small\_fp32 [fp32,imagenet] (exported\_tflite) [gpu-valhall-mali-android31] default-flags,compile-stats](https://perf.iree.dev/serie?IREE?954bc3dc1fd0c22768ebfe898a67c0db3743d74e8fb776fced75eafb0421058f-e54cd682-c079-4c42-b4ad-d92c4bedea13)	29133 (vs. 25483, 14.32%↑)
> [BertForMaskedLMTF [fp32,seqlen512,tensorflow] (exported\_tf\_v2) [cpu-x86\_64-cascadelake-linux-gnu] experimental-flags,fuse-padding,compile-stats](https://perf.iree.dev/serie?IREE?e16d3f99f851c11fef6be64c7f06a637b410618f2618cf16aa599b54ea8970e3-e54cd682-c079-4c42-b4ad-d92c4bedea13)	16692 (vs. 14638, 14.03%↑)
> [DeepLabV3\_fp32 [fp32] (exported\_tflite) [gpu-adreno-generic-android31] experimental-flags,fuse-padding,compile-stats](https://perf.iree.dev/serie?IREE?0d524f6ce80da6d1998eb8978623a2f6efd413e0b973c6f2dddf52a46b19907e-e54cd682-c079-4c42-b4ad-d92c4bedea13)	16374 (vs. 14458, 13.25%↑)
> _[Top 3 out of 21 results showed]_
> 
> ### Improved Compilation Times 🎉
> Benchmark Name	Compilation Time (ms)
> [DeepLabV3\_fp32 [fp32] (exported\_tflite) [gpu-valhall-mali-android31] experimental-flags,fuse-padding,compile-stats](https://perf.iree.dev/serie?IREE?7bec578c7016cb7e017057c227a9b677901d14c0fff35e31c4a5cf12692db105-e54cd682-c079-4c42-b4ad-d92c4bedea13)	12921 (vs. 15223, 15.12%↓)
> [MiniLML12H384Uncased [int32,seqlen128] (exported\_tf\_v2) [cpu-x86\_64-cascadelake-linux-gnu] experimental-flags,fuse-padding,compile-stats](https://perf.iree.dev/serie?IREE?838cc09b422958a332fd76cf12a6a2a95b8346c8e8d2fe7b15cb5ace4c20581e-e54cd682-c079-4c42-b4ad-d92c4bedea13)	14380 (vs. 16497, 12.83%↓)
> [DeepLabV3\_fp32 [fp32] (exported\_tflite) [gpu-valhall-mali-android31] default-flags,compile-stats](https://perf.iree.dev/serie?IREE?7731488e1eb90da5480e76b4cd98b12c16b83d7c7011b0aa9ef3a5d6d2059a3c-e54cd682-c079-4c42-b4ad-d92c4bedea13)	13518 (vs. 15338, 11.87%↓)
> _[Top 3 out of 14 results showed]_
> 
> For more information:
> 
> * [Full benchmark result tables](https://gist.github.com/47d3691008b0ed4489a33cdf6c472cab)
> * [Build](https://github.com/iree-org/iree/actions/runs/4206614131/attempts/1)
> 
> [Source Workflow Run](https://github.com/iree-org/iree/actions/runs/4206614131)

Now I see these results for the x86_64 benchmarks. I didn't check it carefully before the merge, so that's an error we need to avoid.

But also the x86_64 benchmark results there are false positives... Because we don't see the regressions in the post submit results
image

The Android benchmarks have the true regressions, but we didn't run them in the presubmit for this PR.

It seems like the fix here is tuning down the noises from the benchmark machines, blocking the merge if we detect the regressions, and providing a way to manually confirm and log the justification. We should include these in the following plans for benchmark hardening. cc @julianwa

**Comment by GMNGeoffrey** [comment]
I think we should also probably make the flag-file parsing not whitespace sensitive...

**Comment by pzread** [comment]
> I think we should also probably make the flag-file parsing not whitespace sensitive...

Took a second look, my memory was wrong. The wrong flagfile is:
```
--device=local-task
--module=../../vmfb/mobilenet_v2_1.0_224.tflite.mlir-b94b73cfb69caaf1f433f3334c01e67d.vmfb
--function=main
--input=1x224x224x3xf32
--task_topology_group_count=4 --device_allocator=caching
```
It was two flags at the same line, which is invalid for flagfile I think

**Comment by GMNGeoffrey** [comment]
Ah it seems more reasonable to disallow that, but probably should fail flag parsing or similar? Either that's a valid setting of device_allocator, it's a parsing failure, or it's an invalid value for task_topology_group_count. Silent failure seems bad

**Pull request merged**
